### PR TITLE
[Shipping labels] Upsert packages response after fetching from remote

### PR DESCRIPTION
**Pull request by rachelmcr**
Part of: #14522
⚠️ Depends on #14688 ⚠️ 

## Description

This PR caches the packages response so we can fetch packages from storage during Woo Shipping label creation:

* Adds `ReadOnlyConvertible` extensions to each storage entity.
* Adds caching to `WooShippingStore`:
   * Adds helpers to upsert the packages response in storage.
   * Updates `loadPackages(siteID:completion:)` to upsert the response from remote.

Note: I considered splitting up this PR, but the `ReadOnlyConvertible` extensions are in support of the changes to `WooShippingStore`, and all of those changes are tested collectively in the new unit test in `WooShippingStoreTests`.

## Testing information

Confirm unit tests pass. You can also confirm the packages response continues to work as expected:

1. Install and set up the Woo Shipping extension on your store.
2. Build and run the app with the revampedShippingLabelCreation feature flag enabled.
3. Create an order with the processing status and at least one physical product.
4. In your WooCommerce product settings on the web, change the weight and dimension units and save your changes.
5. On the Orders tab in the app, open the order you created.
6. In the order details, select "Create Shipping Label."
7. Immediately tap "Select a Package."
8. Select each tab and confirm the expected packages appear.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Yosemite/Yosemite/Model/Storage/WooShippingCarrierPredefinedOptions+ReadOnlyConvertible.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    }
+}
+
+// Storage.WooShippingCarrierPredefinedOptions: ReadOnlyConvertible Conformance.
+//
+extension Storage.WooShippingCarrierPredefinedOptions: ReadOnlyConvertible {
+    /// Updates the Storage.WooShippingCarrierPredefinedOptions with the a ReadOnly WooShippingCarrierPredefinedOptions.
+    ///
+    public func update(with carrierPredefinedOptions: Yosemite.WooShippingCarrierPredefinedOptions) {
+        self.carrierID = carrierPredefinedOptions.carrierID
```
Comment (bozidarsevo) [inline Yosemite/Yosemite/Model/Storage/WooShippingCarrierPredefinedOptions+ReadOnlyConvertible.swift:16]: Should we also set `predefinedOptions` ?

File: Yosemite/Yosemite/Model/Storage/WooShippingPackagesResponse+ReadOnlyConvertible.swift [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    }
+}
+
+// Storage.WooShippingPackagesResponse: ReadOnlyConvertible Conformance.
+//
+extension Storage.WooShippingPackagesResponse: ReadOnlyConvertible {
+    /// Updates the Storage.WooShippingPackagesResponse with the a ReadOnly WooShippingPackagesResponse.
+    ///
+    public func update(with packages: Yosemite.WooShippingPackagesResponse) {
+        self.siteID = packages.siteID
```
Comment (bozidarsevo) [inline Yosemite/Yosemite/Model/Storage/WooShippingPackagesResponse+ReadOnlyConvertible.swift:16]: Same as above, shouldn't we also update the rest of the properties?

File: Yosemite/Yosemite/Model/Storage/WooShippingSavedPredefinedPackage+ReadOnlyConvertible.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import Storage
+
+// Storage.WooShippingSavedPredefinedPackage: ReadOnlyConvertible Conformance.
+//
+extension Storage.WooShippingSavedPredefinedPackage: ReadOnlyConvertible {
+    /// Updates the Storage.WooShippingSavedPredefinedPackage with the a ReadOnly WooShippingSavedPredefinedPackage.
+    ///
+    public func update(with savedPackage: Yosemite.WooShippingSavedPredefinedPackage) {
+        self.groupTitle = savedPackage.groupTitle
+        self.providerID = savedPackage.providerID
```
Comment (bozidarsevo) [inline Yosemite/Yosemite/Model/Storage/WooShippingSavedPredefinedPackage+ReadOnlyConvertible.swift:11]: Same as above, shouldn't we also update the rest of the properties?

File: Yosemite/Yosemite/Model/Storage/WooShippingPredefinedOption+ReadOnlyConvertible.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+// Storage.WooShippingPredefinedOption: ReadOnlyConvertible Conformance.
+//
+extension Storage.WooShippingPredefinedOption: ReadOnlyConvertible {
+    /// Updates the Storage.WooShippingPredefinedOption with the a ReadOnly WooShippingPredefinedOption.
+    ///
+    public func update(with option: Yosemite.WooShippingPredefinedOption) {
+        self.title = option.title
+        self.providerID = option.providerID
+
```
Comment (bozidarsevo) [inline Yosemite/Yosemite/Model/Storage/WooShippingPredefinedOption+ReadOnlyConvertible.swift:12]: Same as above, shouldn't we also update the rest of the properties (`predefinedPackages`)?

**Review by bozidarsevo — COMMENTED** [review]
Just few questions in the comments

File: Yosemite/Yosemite/Model/Storage/WooShippingCarrierPredefinedOptions+ReadOnlyConvertible.swift (hunk 1 above)
Comment (rachelmcr) [reply Yosemite/Yosemite/Model/Storage/WooShippingCarrierPredefinedOptions+ReadOnlyConvertible.swift:16]: @bozidarsevo In the `update` methods, we only update the Core Data entity's attributes. In this case, for example, the `Storage.WooShippingCarrierPredefinedOptions` entity only has the attribute `carrierID`.

Other properties in the read-only (Networking) model that are relationships in the Core Data model are stored and updated separately, when we upsert the entities in `WooShippingStore`. We insert those child storage entities, update them, and save the relationship with the parent entity. You can see how that's handled for `WooShippingCarrierPredefinedOptions` and its `WooShippingPredefinedOption`s [in the store here](https://github.com/woocommerce/woocommerce-ios/blob/522db1b99a222ec85f9eaa8217f0c0c78c67f148/Yosemite/Yosemite/Stores/WooShippingStore.swift#L260-L279).

**Review by bozidarsevo — APPROVED** [review]

**Pull request merged**